chained/water_version_control.go

```diff
@@ -0,0 +1,136 @@
+func newWaterVersionControl(dir string) *waterVersionControl {
+	return &waterVersionControl{
+		dir: dir,
+	}
+}
+
+// GetWASM returns the WASM file for the given transport
+// Remember to Close the io.ReadCloser after using it
+func (vc *waterVersionControl) GetWASM(ctx context.Context, transport string, downloader waterWASMDownloader) (io.ReadCloser, error) {
+	path := filepath.Join(vc.dir, transport+".wasm")
+	var f io.ReadCloser
+	f, err := os.Open(path)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, log.Errorf("failed to open file %s: %w", path, err)
+	}
+
+	if os.IsNotExist(err) {
+		f, err = vc.downloadWASM(ctx, transport, downloader)
+		if err != nil {
+			return nil, log.Errorf("failed to download WASM file: %w", err)
+		}
+	}
+
+	if err = vc.markUsed(transport); err != nil {
+		return nil, log.Errorf("failed to update WASM history: %w", err)
+	}
+
+	return f, nil
+}
+
+// Commit will update the history of the last time the WASM file was loaded
+// and delete the outdated WASM files
+func (vc *waterVersionControl) markUsed(transport string) error {
+	f, err := os.Create(filepath.Join(vc.dir, transport+".last-loaded"))
+	if err != nil {
+		return log.Errorf("failed to create file %s: %w", transport+".last-loaded", err)
+	}
+	defer f.Close()
+
+	if _, err = f.WriteString(strconv.FormatInt(time.Now().UTC().Unix(), 10)); err != nil {
+		return log.Errorf("failed to write to file %s: %w", transport+".last-loaded", err)
+	}
+	if err = vc.cleanOutdated(); err != nil {
+		return log.Errorf("failed to clean outdated WASMs: %w", err)
+	}
+	return nil
+}
+
+// unusedWASMsDeletedAfter is the time after which the WASM files are considered outdated
+const unusedWASMsDeletedAfter = 7 * 24 * time.Hour
+
+func (vc *waterVersionControl) cleanOutdated() error {
+	wg := new(sync.WaitGroup)
+	// walk through dir, load last-loaded and delete if older than unusedWASMsDeletedAfter
+	err := filepath.Walk(vc.dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return log.Errorf("failed to walk through dir: %w", err)
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		if filepath.Ext(path) != ".last-loaded" {
+			return nil
+		}
+
+		lastLoaded, err := os.ReadFile(path)
+		if err != nil {
+			return log.Errorf("failed to read file %s: %w", path, err)
+		}
+
+		i, err := strconv.ParseInt(string(lastLoaded), 10, 64)
+		if err != nil {
+			return log.Errorf("failed to parse int: %w", err)
+		}
+		lastLoadedTime := time.Unix(i, 0)
+		if time.Since(lastLoadedTime) > unusedWASMsDeletedAfter {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				transport := strings.TrimSuffix(filepath.Base(path), ".last-loaded")
+				if err = os.Remove(filepath.Join(vc.dir, transport+".wasm")); err != nil {
+					log.Errorf("failed to remove wasm file %s: %w", transport+".wasm", err)
+					return
+				}
+				if err = os.Remove(path); err != nil {
+					log.Errorf("failed to remove last-loaded file %s: %w", path, err)
+					return
+				}
+			}()
```

Comment:
I wonder if this could create issues with the ongoing `filepath.Walk` loop.  You're concurrently deleting files in the same directory you're walking.  I have no idea really.  I guess it depends on what's going on under the hood in `filepath.Walk`.  Might be safer though to do all of the deleting after the `filepath.Walk` call exits.